packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree-widget.tsx

```diff
@@ -0,0 +1,230 @@
+        return <div className='theia-widget-noInfo noEdits'>No edits have been detected in the workspace so far.</div>;
+    }
+
+    protected renderCaption(node: TreeNode, props: NodeProps): React.ReactNode {
+        if (BulkEditInfoNode.is(node)) {
+            return this.decorateBulkEditInfoNode(node);
+        } else if (BulkEditNode.is(node)) {
+            return this.decorateBulkEditNode(node);
+        }
+        return 'caption';
+    }
+
+    protected decorateBulkEditNode(node: BulkEditNode): React.ReactNode {
+        if (node?.parent && node?.bulkEdit && ('edit' in node?.bulkEdit)) {
+            const bulkEdit = node.bulkEdit;
+            const parent = node.parent as BulkEditInfoNode;
+
+            if (parent?.fileContents) {
+                const lines = parent.fileContents.split('\n');
+                const startLineNum = +bulkEdit.edit.range.startLineNumber;
+
+                if (lines.length > startLineNum) {
+                    const startColumn = +bulkEdit.edit.range.startColumn;
+                    const endColumn = +bulkEdit.edit.range.endColumn;
+                    const lineText = lines[startLineNum - 1];
+                    const beforeMatch = (startColumn > 26 ? '... ' : '') + lineText.substr(0, startColumn - 1).substr(-25);
+                    const replacedText = lineText.substring(startColumn - 1, endColumn - 1);
+                    const afterMatch = lineText.substr(startColumn - 1 + replacedText.length, 75);
+
+                    return <div className='bulkEditNode'>
+                        <div className='message'>
+                            {beforeMatch}
+                            <span className="replaced-text">{replacedText}</span>
+                            <span className="inserted-text">{bulkEdit.edit.text}</span>
+                            {afterMatch}
+                        </div>
+                    </div>;
+                }
+            }
+        }
+    }
+
+    protected decorateBulkEditInfoNode(node: BulkEditInfoNode): React.ReactNode {
+        const icon = this.toNodeIcon(node);
+        const name = this.toNodeName(node);
+        const description = this.toNodeDescription(node);
+        const path = this.labelProvider.getLongName(node.uri.withScheme('bulkedit'));
+        return <div title={path} className='bulkEditInfoNode'>
+            {icon && <div className={icon + ' file-icon'}></div>}
+            <div className='name'>{name}</div>
+            <div className='path'>{description}</div>
+        </div>;
+    }
+
+    private async getFileContentsMap(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<Map<string, string>> {
+        const fileContentMap = new Map<string, string>();
+
+        if (workspaceEdit?.edits) {
+            for (const element of workspaceEdit.edits) {
+                if (element) {
+                    const filePath = (('newUri' in element) && element?.newUri?.path) || (('resource' in element) && element?.resource?.path);
```

Comment:
do you need `('newUri' in element)` and `('resource' in element)` conditions?

Reply:
yes.  A workspaceEdit can be Text of File and each has different props and therefore the check is needed
![image](https://user-images.githubusercontent.com/71069170/105686908-7c00d480-5f00-11eb-9fed-11a101f8d0b5.png)